File: ios/WemapLivemapModule.m
//
//  WemapLivemapModule.m
//  react-native-wemap-livemap
//
//  Created by Thibault Capelli on 07/04/2022.
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(WemapLivemapModule, NSObject)
RCT_EXTERN_METHOD(
                  drawPolylineViaModule: (NSArray *)coordinatesList
                  options: (NSDictionary *)options
                  resolve: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(
                  getUserLocationModule: (RCTPromiseResolveBlock)resolve
                  reject: (RCTPromiseRejectBlock)reject
)
@end

File: ios/WemapLivemapModule.swift
//
//  WemapLivemapModule.swift
//  react-native-wemap-livemap
//
//  Created by Thibault Capelli on 07/04/2022.
//

import Foundation
import livemap_ios_sdk

@objc(WemapLivemapModule)
class WemapLivemapModule: NSObject {
    var livemapView: WemapLivemap? = nil
    
    @objc static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    internal func setLivemapView(livemapView: WemapLivemap) {
        self.livemapView = livemapView
    }
    
    @available(iOS 14.0, *)
    @objc func drawPolylineViaModule(
        _ coordinatesList: NSArray,
        options: NSDictionary? = nil,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: RCTPromiseRejectBlock
    ) -> Void {
        DispatchQueue.main.async {
            self.livemapView?.wemap.drawPolyline(
                coordinatesList: coordinatesList.map( { Coordinates.fromDictionary($0 as! NSDictionary) } ),
                options: options != nil ? PolylineOptions.fromDictionary(options!) : nil,
                completion: { id in resolve(id) }
            )
        }
    }
    
    @available(iOS 14.0, *)
    @objc func getUserLocationModule(_ resolve: @escaping RCTPromiseResolveBlock,
                                     reject: RCTPromiseRejectBlock
    ) -> Void {
        func resolveLocationDict(coords: Coordinates){
            let dict = NSDictionary(dictionary:
                [
                    "latitude": coords.latitude,
                    "longitude": coords.longitude,
                    "altitude": coords.altitude,
                    "accuracy": coords.accuracy
                ]
            )
            resolve(dict)
        }
        self.livemapView?.wemap.getUserLocation(
            completionHandler: { coordinates in
                resolveLocationDict(coords: coordinates)
            }
        )

    }

    
}
